Enno focus metavariables (#714)
returntocorp/semgrep-docs: docs/writing-rules/rule-syntax.md

In short, `focus-metavariable: $X` is not a pattern in itself, it does not perform any matching, it only focuses the matching on the code already bound to `$X` by other patterns. Whereas `pattern: $X` matches `$X` against your code (and in this context, `$X` only matches expressions)!

> You can also focus multiple metavariables at once by specifying them in a list within a `focus-metavariable`. This will highlight all of the regions that are captured by the metavariables being focused.

Author: LGTM!

<iframe src="https://semgrep.dev/embed/editor?snippet=493E" border="0" frameBorder="0" width="100%" height="432"></iframe>

> However, note that this is different than using multiple `focus-metavariable`s. Because `focus-metavariable` _has_ to be directly under a `patterns`, using multiple `focus-metavariable`s will cause Semgrep to find the overlapping region of all of the focused code (see example below).

Reviewer: @enncoded This contradicts what you wrote above: `You can also focus multiple metavariables at once` - so are you trying to say that there are two different ways how to introduce multiple focus metavariables? Both examples you included provide the same rule, only the formatting of the test samples is different.

You probably mean something about the difference between:
```
    patterns:
      - pattern: foo($X, ..., $Y)
      - focus-metavariable: 
        - $X
        - $Y
```

and:

```
    patterns:
      - pattern: foo($X, ..., $Y)
      - focus-metavariable: $X
      - focus-metavariable: $Y
```

But which of them is which? Both links lead to the same-looking rule (two times `focus-metavariable` key, not a list under one `focus-metavariable` key). Which one is correct syntax, please?

```suggestion
However, note that this is different than using multiple focus metavariables. Include the `focus-metavariable` directly under a `patterns`, using multiple `focus-metavariable`s causes Semgrep to find the overlapping region of all of the focused code (see example below).
```

Author: Looks good!

One little nitpick though, could we add a word or two that convey a sense of exclusivity for line 171? Something like, "Semgrep _only_ finds the overlapping region of all the focused code"

Also, should we use this editor snippet example under that section instead of the current one? I think it does a better job at demonstrating the overlapping property: https://semgrep.dev/s/e8OL. We could also say something like "you can try removing the `focus-metavariable`s to see how they restrict the matching".

Let me know what you think!

Reviewer: @enncoded Thanks for the heads up. Line 171 in the suggestion or code? :D Do you mean adding a word about the exclusivity of results to:

```
Including more `focus-metavariable` keys with different metavariables under the `pattern` key lets Semgrep find the overlapping region of all the focused code:
```

Updating in the suggestion below if you meant that. :-) I also changed the link as you requested, I like the test code in the new example! To avoid having us to move this to different sections, I marked the particular examples with Recommended and Advanced.

Please, can you also send me a link to an example of the list of metavariables? I mean Playground example that you find helpful to demonstrate:

```yaml
    patterns:
      - pattern: foo($X, ..., $Y)
      - focus-metavariable: 
        - $X
        - $Y
```

Suggestion:

```suggestion
There are two ways how to use multiple metavariables. Each of them can lead to different results:

- **Recommended**: Including `focus-metavariable` with more metavariables directly under the `patterns` key returns results regardless of whether there is an overlap between these variables:
    ```yaml
        patterns:
          - pattern: foo($X, ..., $Y)
          - focus-metavariable: 
            - $X
            - $Y
    ```
- **Advanced**: Including more `focus-metavariable` keys with different metavariables under the `pattern` key lets Semgrep find **only** the overlapping region of all the focused code:
    ```yaml
        patterns:
          - pattern: foo($X, ..., $Y)
          - focus-metavariable: $X
          - focus-metavariable: $Y
    ```
    See also the following example:
    <iframe src="https://semgrep.dev/embed/editor?snippet=e8OL" border="0" frameBorder="0" width="100%" height="432"></iframe>
```

Author: > There are two ways how to use multiple metavariables

I assume "how" is a typo here?

> Updating in the suggestion below if you meant that

Yep! You interpreted that correctly, sorry about the confusion.

> Please, can you also send me a link to an example of the list of metavariables?

I'll add an embedded example for the "Recommended" case when the new Semgrep release comes out.

Another small thing, could we change the word "Recommended" to "Common Case" (or similar)? I think "Recommended" implies that the other one is wrong, which in some cases it might not be.

Reviewer: @enncoded thank you! I think that "how" was ok in this case, but I used a different syntax here "There are two methods of using multiple metavariables."

I left "TODO IFRAME" for us to later add the Playground example. :-)

I am not sure what would be better than "Recommended" in our case - I do not think we are saying that the other example is wrong - we only say that the other method is "Advanced". Looking at [Merrian Webster's synonyms](https://www.merriam-webster.com/thesaurus/recommended), I am not sure really. "Common" does not catch what we are trying to say here ("Use mostly this method, do not be afraid to try the other also").

> <iframe src="https://semgrep.dev/embed/editor?snippet=19wx" border="0" frameBorder="0" width="100%" height="432"></iframe>

Reviewer: Note: Not working still at the time of review.

Use multiple focus metavariables in one Semgrep rule by specifying each under a new `focus-metavariable` item in a YAML list. With this rule, Semgrep matches all pieces of code captured by the focus metavariables.

<iframe src="https://semgrep.dev/embed/editor?snippet=493E" border="0" frameBorder="0" width="100%" height="432"></iframe>

There are two methods of using multiple metavariables. Each of them can lead to different results:

- **Recommended**: Including `focus-metavariable` with more metavariables directly under the `patterns` key returns results regardless of whether there is an overlap between these variables:
    ```yaml
        patterns:
          - pattern: foo($X, ..., $Y)
          - focus-metavariable: 
            - $X
            - $Y
    ```
    See the following example:
    TODO IFRAME 
- **Advanced**: Including more `focus-metavariable` keys with different metavariables under the `pattern` key lets Semgrep find **only** the overlapping region of all the focused code:
    ```yaml
        patterns:
          - pattern: foo($X, ..., $Y)
          - focus-metavariable: $X
          - focus-metavariable: $Y
    ```
    See the following example:

>     <iframe src="https://semgrep.dev/embed/editor?snippet=e8OL" border="0" frameBorder="0" width="100%" height="432"></iframe>

Reviewer: @enncoded thank you! I think that "how" was ok in this case, but I used a different syntax here "There are two methods of using multiple metavariables."

I left "TODO IFRAME" for us to later add the Playground example. :-)

I am not sure what would be better than "Recommended" in our case - I do not think we are saying that the other example is wrong - we only say that the other method is "Advanced". Looking at [Merrian Webster's synonyms](https://www.merriam-webster.com/thesaurus/recommended), I am not sure really. "Common" does not catch what we are trying to say here ("Use mostly this method, do not be afraid to try the other also").

Reviewer: Maybe we can say "Standard" instead of "Recommended" but it suggest that "Standard" is somehow lower quality than "Advanced", which is not what we want.

Author: Hm makes sense. Yeah we can just leave it as it is now.
